File: stock_platform/java/backend/java/src/main/java/com/stockplatform/service/PositionService.java
package com.stockplatform.service;

import com.my.mybatis.flex.core.condition.QueryCondition;
import com.stockplatform.dto.PositionCreateDTO;
import com.stockplatform.dto.PositionResponseDTO;
import com.stockplatform.entity.Position;
import com.stockplatform.entity.StockPrice;
import com.stockplatform.mapper.PositionMapper;
import com.stockplatform.mapper.StockPriceMapper;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class PositionService {

    private final PositionMapper positionMapper;
    private final StockPriceMapper stockPriceMapper;

    public PositionService(PositionMapper positionMapper, StockPriceMapper stockPriceMapper) {
        this.positionMapper = positionMapper;
        this.stockPriceMapper = stockPriceMapper;
    }

    public PositionResponseDTO createPosition(PositionCreateDTO dto) {
        Position position = new Position();
        position.setAccountId(dto.getAccountId());
        position.setStockCode(dto.getStockCode());
        position.setStockName(dto.getStockName());
        position.setQuantity(dto.getQuantity());
        position.setAvgCost(dto.getAvgCost());
        position.setCreatedAt(LocalDateTime.now());
        position.setUpdatedAt(LocalDateTime.now());
        positionMapper.insertSelective(position);
        return enrichPositionWithPrice(PositionResponseDTO.fromEntity(position));
    }

    public List<PositionResponseDTO> listPositions(Integer accountId) {
        QueryCondition condition = QueryCondition.create();
        if (accountId != null) {
            condition.where("account_id = ?", accountId);
        }
        return positionMapper.selectAllByCondition(condition).stream()
                .map(PositionResponseDTO::fromEntity)
                .map(this::enrichPositionWithPrice)
                .collect(Collectors.toList());
    }

    public List<PositionResponseDTO> listPositionsByAccount(Integer accountId) {
        QueryCondition condition = QueryCondition.create()
                .where("account_id = ?", accountId);
        return positionMapper.selectAllByCondition(condition).stream()
                .map(PositionResponseDTO::fromEntity)
                .map(this::enrichPositionWithPrice)
                .collect(Collectors.toList());
    }

    public PositionResponseDTO getPosition(Integer id) {
        Position position = positionMapper.selectOneById(id);
        if (position == null) {
            throw new RuntimeException("持仓不存在");
        }
        return enrichPositionWithPrice(PositionResponseDTO.fromEntity(position));
    }

    public PositionResponseDTO updatePosition(Integer id, PositionCreateDTO dto) {
        Position position = positionMapper.selectOneById(id);
        if (position == null) {
            throw new RuntimeException("持仓不存在");
        }

        int totalQuantity = position.getQuantity() + dto.getQuantity();
        if (totalQuantity > 0) {
            double totalCost = position.getQuantity() * position.getAvgCost()
                    + dto.getQuantity() * dto.getAvgCost();
            position.setAvgCost(totalCost / totalQuantity);
        }
        position.setQuantity(totalQuantity);
        position.setUpdatedAt(LocalDateTime.now());
        positionMapper.update(position);
        return enrichPositionWithPrice(PositionResponseDTO.fromEntity(position));
    }

    public void deletePosition(Integer id) {
        positionMapper.deleteById(id);
    }

    public long getPositionCount() {
        return positionMapper.selectCount(null);
    }

    private PositionResponseDTO enrichPositionWithPrice(PositionResponseDTO dto) {
        QueryCondition condition = QueryCondition.create()
                .where("stock_code = ?", dto.getStockCode())
                .orderBy("trade_date", true);
        List<StockPrice> prices = stockPriceMapper.selectAllByCondition(condition);
        if (!prices.isEmpty()) {
            StockPrice latest = prices.get(0);
            dto.setCurrentPrice(latest.getPrice());
            dto.setChangeValue(latest.getChangeValue());
            dto.setChangePercent(latest.getChangePercent());
            dto.setMarketValue(dto.getQuantity() * latest.getPrice());
        }
        return dto;
    }
}

File: stock_platform/python/backend/python/services/position_service.py
from models import Position
from schemas import PositionCreate
from datetime import datetime
from tortoise.exceptions import DoesNotExist


async def create_position(position: PositionCreate) -> Position:
    return await Position.create(
        account_id=position.account_id,
        stock_code=position.stock_code,
        stock_name=position.stock_name,
        quantity=position.quantity,
        avg_cost=position.avg_cost
    )


async def list_positions(account_id: int | None = None) -> list[Position]:
    query = Position.all()
    if account_id:
        query = query.filter(account_id=account_id)
    return await query


async def list_positions_by_account(account_id: int) -> list[Position]:
    return await Position.filter(account_id=account_id).all()


async def get_position(position_id: int) -> Position | None:
    try:
        return await Position.get(id=position_id)
    except DoesNotExist:
        return None


async def update_position(position_id: int, position_data: PositionCreate) -> Position | None:
    position = await get_position(position_id)
    if not position:
        return None

    total_quantity = position.quantity + position_data.quantity
    if total_quantity > 0:
        total_cost = position.quantity * position.avg_cost + position_data.quantity * position_data.avg_cost
        position.avg_cost = total_cost / total_quantity
    position.quantity = total_quantity
    position.updated_at = datetime.now()

    await position.save()
    return position


async def delete_position(position_id: int) -> bool:
    position = await get_position(position_id)
    if not position:
        return False
    await position.delete()
    return True